src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,211 @@
+            } else {
+                newPrev.next = newNode;
+            }
+            newPrev = newNode;
+        }
+        if (newNext == null) {
+            last = newPrev;
+        } else {
+            newPrev.next = newNext;
+            newNext.prev = newPrev;
+        }
+        size += lengthArrayCollection;
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> setNode = node(index);
+        T oldVal = setNode.item;
+        setNode.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
+    T unlink(Node<T> x) {
+        final T element = x.item;
+        final Node<T> next = x.next;
+        final Node<T> prev = x.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+        x.item = null;
+        size--;
+        return element;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> newFirst = first;
+            for (int i = 0; i < index; i++) {
+                newFirst = newFirst.next;
+            }
+            return newFirst;
+        } else {
+            Node<T> newLast = last;
+            for (int i = size - 1; i > index; i--) {
+                newLast = newLast.prev;
+            }
+            return newLast;
+        }
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
```

Comment:
Not informative message
```suggestion
        return "Invalid index: " + index + " for size: " + size;
```